Build failure: pkgsLLVM.x265

### Steps To Reproduce

Steps to reproduce the behavior:
1. build `pkgsLLVM.x265`

### Build log

```
make[2]: *** [CMakeFiles/x265-shared.dir/build.make:90: ipfilter8.S.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:227: CMakeFiles/x265-shared.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
<instantiation>:2:9: error: unknown directive
        .endfunc
        ^
<instantiation>:32:1: note: while in macro instantiation
endfunc
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:412:1: note: while in macro instantiation
LUMA_HPS 64 32
^
<instantiation>:11:9: error: unknown directive
        .func x265_interp_8tap_horiz_ps_64x48_neon
        ^
<instantiation>:1:1: note: while in macro instantiation
function x265_interp_8tap_horiz_ps_64x48_neon
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:413:1: note: while in macro instantiation
LUMA_HPS 64 48
^
<instantiation>:2:9: error: unknown directive
        .endfunc
        ^
<instantiation>:32:1: note: while in macro instantiation
endfunc
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:413:1: note: while in macro instantiation
LUMA_HPS 64 48
^
<instantiation>:11:9: error: unknown directive
        .func x265_interp_8tap_horiz_ps_64x64_neon
        ^
<instantiation>:1:1: note: while in macro instantiation
function x265_interp_8tap_horiz_ps_64x64_neon
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:414:1: note: while in macro instantiation
LUMA_HPS 64 64
^
<instantiation>:2:9: error: unknown directive
        .endfunc
        ^
<instantiation>:32:1: note: while in macro instantiation
endfunc
^
/build/x265_3.5/source/common/aarch64/ipfilter8.S:414:1: note: while in macro instantiation
LUMA_HPS 64 64
^
make[2]: *** [CMakeFiles/x265-static.dir/build.make:90: ipfilter8.S.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:147: CMakeFiles/x265-static.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

### Additional context

Add any other context about the problem here.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@codyopel 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/x265/default.nix
+++ b/pkgs/development/libraries/x265/default.nix
@@ -1,4 +1,5 @@
 { lib
+, gccStdenv
 , stdenv
 , fetchurl
 , fetchpatch
@@ -66,7 +67,11 @@ stdenv.mkDerivation rec {
     })
     # Fix detection of NEON (and armv6 build) :
     ./fix-neon-detection.patch
-  ];
+  ]
+    # CMake files require a bit of patching to support CMAKE_ASM_COMPILER.
+    # Made by @RossComputerGuy for x265 v3.5.
+    # https://mailman.videolan.org/pipermail/x265-devel/2024-July/013734.html
+    ++ lib.optional (stdenv.cc.isClang && !stdenv.targetPlatform.isDarwin) ./fix-clang-asm.patch;
 
   postPatch = ''
     substituteInPlace cmake/Version.cmake \
@@ -89,7 +94,9 @@ stdenv.mkDerivation rec {
     (mkFlag ppaSupport "ENABLE_PPA")
     (mkFlag vtuneSupport "ENABLE_VTUNE")
     (mkFlag werrorSupport "WARNINGS_AS_ERRORS")
-  ];
+  ]
+    # Clang does not support the endfunc directive so use GCC.
+    ++ lib.optional (stdenv.cc.isClang && !stdenv.targetPlatform.isDarwin) "-DCMAKE_ASM_COMPILER=${gccStdenv.cc}/bin/${gccStdenv.cc.targetPrefix}gcc";
 
   cmakeStaticLibFlags = [
     "-DHIGH_BIT_DEPTH=ON"

--- a/pkgs/development/libraries/x265/fix-clang-asm.patch
+++ b/pkgs/development/libraries/x265/fix-clang-asm.patch
@@ -1,0 +1,34 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index a407271b4..593628e0f 100755
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -310,7 +310,7 @@ if(GCC)
+     endif()
+     check_cxx_compiler_flag(-mstackrealign CC_HAS_STACK_REALIGN) 
+     if (CC_HAS_STACK_REALIGN)
+-        add_definitions(-mstackrealign)
++        add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-mstackrealign>)
+     endif()
+     # Disable exceptions. Reduce executable size, increase compability.
+     check_cxx_compiler_flag(-fno-exceptions CC_HAS_FNO_EXCEPTIONS_FLAG)
+@@ -545,7 +545,7 @@ if((MSVC_IDE OR XCODE OR GCC) AND ENABLE_ASSEMBLY)
+             list(APPEND ASM_OBJS ${ASM}.${SUFFIX})
+             add_custom_command(
+                 OUTPUT ${ASM}.${SUFFIX}
+-                COMMAND ${CMAKE_CXX_COMPILER}
++                COMMAND ${CMAKE_ASM_COMPILER}
+                 ARGS ${ARM_ARGS} -c ${ASM_SRC} -o ${ASM}.${SUFFIX}
+                 DEPENDS ${ASM_SRC})
+         endforeach()
+diff --git a/common/CMakeLists.txt b/common/CMakeLists.txt
+index 12b643ad5..876f3543d 100644
+--- a/common/CMakeLists.txt
++++ b/common/CMakeLists.txt
+@@ -16,6 +16,7 @@ endif(EXTRA_LIB)
+ if(ENABLE_ASSEMBLY)
+     set_source_files_properties(threading.cpp primitives.cpp pixel.cpp PROPERTIES COMPILE_FLAGS -DENABLE_ASSEMBLY=1)
+     list(APPEND VFLAGS "-DENABLE_ASSEMBLY=1")
++    enable_language(ASM)
+ endif(ENABLE_ASSEMBLY)
+ 
+ if(ENABLE_ASSEMBLY AND X86)

